File: prolog/p-37.pl
:- ensure_loaded('p-36').

transform(P-M, Acc, Res) :-
    P1 is P - 1, M1 is M - 1,
    Res is Acc * P1 * P1 ^ M1.

phi(N, Ans) :-
    prime_factors_mult(N, Xs),
    foldl(transform, Xs, 1, Ans).

File: erlang/p_37.erl
-module(p_37).
-export([phi/1]).
-import(p_36, [prime_factors_mult/1]).

pow(_, 0) ->
    1;
pow(A, 1) ->
    A;
pow(A, B) when B rem 2 == 0 ->
    Temp = pow(A, B div 2),
    Temp * Temp;
pow(A, B) ->
    Temp = pow(A, B div 2),
    Temp * Temp * A.

phi(N) ->
    Xs = prime_factors_mult(N),
    lists:foldl(fun ({P, M}, Acc) ->
			Acc * (P - 1) * pow(P, M - 1)
		end, 1, Xs).
